crates/rpc/src/supervisor.rs

```diff
@@ -0,0 +1,145 @@
+///
+/// <https://github.com/ethereum-optimism/optimism/blob/develop/op-supervisor/supervisor/frontend/frontend.go#L18-L28>
+#[async_trait]
+pub trait Supervisor {
+    /// The error returned by supervisor methods.
+    type Error;
+
+    /// Returns the safety level of a single message.
+    async fn check_message(
+        &self,
+        identifier: MessageIdentifier,
+        payload_hash: B256,
+    ) -> Result<SafetyLevel, Self::Error>;
+
+    /// Returns if the messages meet the minimum safety level.
+    async fn check_messages(
+        &self,
+        messages: &[ExecutingMessage],
+        min_safety: SafetyLevel,
+    ) -> Result<(), Self::Error>;
+
+    /// Returns the block at which the given block was derived from.
+    async fn cross_derived_from(
+        &self,
+        chain_id: u32,
+        derived: BlockNumHash,
+    ) -> Result<BlockInfo, Self::Error>;
+
+    /// Returns the local unsafe block.
+    async fn local_unsafe(&self, chain_id: u32) -> Result<BlockNumHash, Self::Error>;
+
+    /// Returns the cross safe block.
+    async fn cross_safe(&self, chain_id: u32) -> Result<DerivedIdPair, Self::Error>;
+
+    /// Returns the finalized block.
+    async fn finalized(&self, chain_id: u32) -> Result<BlockNumHash, Self::Error>;
+
+    /// Returns the finalized L1 block.
+    async fn finalized_l1(&self) -> Result<BlockInfo, Self::Error>;
+
+    /// Returns the super root at the given timestamp.
+    async fn super_root_at_timestamp(
+        &self,
+        timestamp: u64,
+    ) -> Result<SuperRootResponse, Self::Error>;
+
+    /// Returns all safe derived blocks at the given block.
+    async fn all_safe_derived_at(
+        &self,
+        derived_from: BlockNumHash,
+    ) -> Result<Vec<(u32, BlockNumHash)>, Self::Error>;
+}
+
+/// An error from the `op-supervisor`.
+#[derive(Copy, Clone, Debug, Display, From)]
+pub enum SupervisorError {
+    /// A failed request.
+    RequestFailed,
+}
+
+/// A supervisor client.
+#[derive(Debug, Clone)]
+pub struct SupervisorClient {
+    /// The inner RPC client.
+    client: ReqwestClient,
+}
+
+impl SupervisorClient {
+    /// Creates a new `SupervisorClient` with the given `client`.
+    pub const fn new(client: ReqwestClient) -> Self {
+        Self { client }
+    }
+}
+
+#[async_trait]
+impl Supervisor for SupervisorClient {
+    type Error = SupervisorError;
+
+    async fn check_message(
+        &self,
+        identifier: MessageIdentifier,
+        payload_hash: B256,
+    ) -> Result<SafetyLevel, Self::Error> {
+        self.client
+            .request("supervisor_checkMessage", (identifier, payload_hash))
+            .await
+            .map_err(|_| SupervisorError::RequestFailed)
+    }
+
+    async fn check_messages(
+        &self,
+        messages: &[ExecutingMessage],
+        min_safety: SafetyLevel,
+    ) -> Result<(), Self::Error> {
+        self.client
+            .request("supervisor_checkMessages", (messages, min_safety))
```

Comment:
I think we actually want an array here (unless alloy has gotten smarter). Does this already work?

Reply:
Yeah, reading the go impl, the call will end up marshaling arguments as an array [here](https://github.com/ethereum-optimism/op-geth/blob/optimism/rpc/client.go#L552-L555).

Reply:
It must be tuple, you can't have an array of differently typed items in rust

Reply:
Yeah, ik - you can if they're encoded to hex (both `String`)